crates/rattler_networking/src/authentication_storage/fallback_storage.rs

```diff
@@ -30,44 +31,68 @@ pub enum FallbackStorageError {
 impl FallbackStorage {
     /// Create a new fallback storage with the given path
     pub fn new(path: PathBuf) -> Self {
-        Self {
-            path,
-            mutex: Arc::new(Mutex::new(())),
+        Self { path }
+    }
+
+    /// Lock the fallback storage file for reading and writing. This will block until the lock is
+    /// acquired.
+    fn lock(&self) -> Result<LockFile, FallbackStorageError> {
+        std::fs::create_dir_all(self.path.parent().unwrap())?;
+        let path = self.path.with_extension("lock");
+        let mut lock = fslock::LockFile::open(&path).map_err(|e| {
+            FallbackStorageError::FailedToLock(path.to_string_lossy().into_owned(), e)
+        })?;
+
+        // First try to lock the file without block. If we can't immediately get the lock we block and issue a debug message.
+        if !lock.try_lock_with_pid().map_err(|e| {
+            FallbackStorageError::FailedToLock(path.to_string_lossy().into_owned(), e)
+        })? {
+            tracing::debug!("waiting for lock on {}", path.to_string_lossy());
+            lock.lock_with_pid().map_err(|e| {
+                FallbackStorageError::FailedToLock(path.to_string_lossy().into_owned(), e)
+            })?;
         }
+
+        Ok(lock)
     }
 
     /// Store the given authentication information for the given host
     pub fn set_password(&self, host: &str, password: &str) -> Result<(), FallbackStorageError> {
-        let _lock = self.mutex.lock().unwrap();
+        let _lock = self.lock()?;
         let mut dict = self.read_json()?;
         dict.insert(host.to_string(), password.to_string());
         self.write_json(&dict)
     }
 
     /// Retrieve the authentication information for the given host
     pub fn get_password(&self, host: &str) -> Result<Option<String>, FallbackStorageError> {
-        let _lock = self.mutex.lock().unwrap();
+        let _lock = self.lock()?;
         let dict = self.read_json()?;
         Ok(dict.get(host).cloned())
     }
 
     /// Delete the authentication information for the given host
     pub fn delete_password(&self, host: &str) -> Result<(), FallbackStorageError> {
-        let _lock = self.mutex.lock().unwrap();
+        let _lock = self.lock()?;
         let mut dict = self.read_json()?;
         dict.remove(host);
         self.write_json(&dict)
     }
 
     /// Read the JSON file and deserialize it into a HashMap, or return an empty HashMap if the file
     /// does not exist
-    fn read_json(&self) -> Result<std::collections::HashMap<String, String>, FallbackStorageError> {
+    fn read_json(&self) -> Result<HashMap<String, String>, FallbackStorageError> {
         if !self.path.exists() {
-            tracing::warn!(
-                "Can't find path for fallback storage on {}",
-                self.path.to_string_lossy()
-            );
-            return Ok(std::collections::HashMap::new());
+            static WARN_GUARD: Lazy<Mutex<HashSet<PathBuf>>> =
```

Comment:
@wolfv this should ensure that the warning is only printed once per filename for the current process.